cupyx/scipy/sparse/csgraph/_traversal.py

```diff
@@ -0,0 +1,189 @@
+            If ``return_labels`` == ``True``, returns a tuple ``(n, labels)``,
+            where ``n`` is the number of connected components and ``labels`` is
+            labels of each connected components. Otherwise, returns ``n``.
+
+    .. seealso:: :func:`scipy.sparse.csgraph.connected_components`
+    """
+    if connection.lower() not in ('weak', 'strong'):
+        raise ValueError("connection must be 'weak' or 'strong'")
+
+    if connection.lower() == 'weak':
+        directed = False
+
+    if not cupyx.scipy.sparse.isspmatrix_csr(csgraph):
+        csgraph = cupyx.scipy.sparse.csr_matrix(csgraph)
+    util._assert_nd_squareness(csgraph)
+
+    return _connected_components(csgraph, directed, return_labels)
+
+
+def _connected_components(csgraph, directed, return_labels):
+    m = csgraph.shape[0]
+    labels = cupy.arange(m, dtype=numpy.int32)
+    nnz = csgraph.nnz
+    if directed:
+        # Note: The following implementation is naive and may require a lot of
+        # memory to run.
+        data = cupy.ones((nnz,), dtype=numpy.float32)
+        csgraph = cupyx.scipy.sparse.csr_matrix(
+            (data, csgraph.indices, csgraph.indptr), shape=csgraph.shape)
+        csgraph += csgraph * csgraph
+        while csgraph.nnz > nnz:
+            nnz = csgraph.nnz
+            csgraph.data[:] = 1
+            csgraph += csgraph * csgraph
+        _cupy_connect_directed(csgraph.indices, csgraph.indptr, m, labels,
+                               size=nnz)
+    else:
+        _cupy_connect_undirected(csgraph.indices, csgraph.indptr, m, labels,
+                                 size=nnz)
+    count = cupy.zeros((1,), dtype=numpy.int32)
+    root_labels = cupy.empty((m,), dtype=numpy.int32)
+    _cupy_count_components(labels, count, root_labels, size=m)
+    n = int(count[0])
+    if not return_labels:
+        return n
+    _cupy_adjust_labels(n, cupy.sort(root_labels[:n]), labels)
+    return n, labels
+
+
+_GET_ROW_ID_ = '''
+__device__ inline int get_row_id(int i, int min, int max, const int *indptr) {
+    int row = (min + max) / 2;
+    while (min < max) {
+        if (i < indptr[row]) {
+            max = row - 1;
+        } else if (i >= indptr[row + 1]) {
+            min = row + 1;
+        } else {
+            break;
+        }
+        row = (min + max) / 2;
+    }
+    return row;
+}
+'''
+
+_CONNECT_NODES_ = '''
+__device__ inline void connect_nodes(int i, int j, int *labels) {
+    while (true) {
+        while (i != labels[i]) { i = labels[i]; }
+        while (j != labels[j]) { j = labels[j]; }
+        if (i == j) break;
+        if (i < j) {
+            int old = atomicCAS( &labels[j], j, i );
+            if (old == j) break;
+            j = old;
+        } else {
+            int old = atomicCAS( &labels[i], i, j );
+            if (old == i) break;
+            i = old;
+        }
+    }
+}
+'''
+
+_cupy_connect_directed = cupy.core.ElementwiseKernel(
+    'raw I INDICES, raw I INDPTR, int32 M',
```

Comment:
Same as above, let's try to avoid capitalized variable names,
there are some PRs that have been merged with this format but we will be fixing them.

Thanks